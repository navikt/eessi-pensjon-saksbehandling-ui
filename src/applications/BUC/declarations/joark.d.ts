
export interface JoarkFileVariant {
  variantformat:string;
  filnavn: string;
}

export interface JoarkDoc {
  dokumentInfoId: string;
  tittel: string;
  dokumentvarianter: Array<JoarkFileVariant>;
}
export interface JoarkPoster {
  journalpostId: string;
  tittel: string;
  tema: string;
  datoOpprettet: string;
  tilleggsopplysninger?: string;
  dokumenter: Array<JoarkDoc>
}

export interface JoarkFile {
  journalpostId: string;
  tittel: string;
  tema: string;
  datoOpprettet: Date;
  dokumentInfoId: string;
  tilleggsopplysninger?: string;
  variant: JoarkFileVariant
}

interface Content {
  base64: string;
}

export interface JoarkFileWithContent extends JoarkFile {
  content: Content;
  name: string;
  size: number;
  mimetype: string;
}
