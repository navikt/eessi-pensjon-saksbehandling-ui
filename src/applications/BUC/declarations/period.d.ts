export interface PeriodDate {
  day: string;
  month: string;
  year: string;
}

export interface Country {
  label: string;
  value: string;
}

export interface Period {
  id?: number,
  type: string;
  startDate: PeriodDate;
  endDate?: PeriodDate;
  dateType: string;
  uncertainDate: boolean;
  country: Country;
  comment: string;
  otherType?: string;
  insuranceName?: string;
  insuranceType?: string;
  insuranceId ?: string;
  workActivity ?: string;
  workName ?: string;
  workType ?: string;
  workStreet ?: string;
  workCity ?: string;
  workZipCode ?: string;
  workRegion ?: string;
  childFirstName ?: string;
  childLastName ?: string;
  childBirthDate ?: PeriodDate;
  learnInstitution ?: string;
  payingInstitution ?: string;
  attachments ?: Array[any];
  [key: string]: any;
}

export type Periods = Array<Period>

export type PeriodErrors = {[k: string]: string | undefined}

export interface PayloadPeriod {
  land: string;
  periode: null | {
    lukketPeriode?: {
      fom: string | null;
      tom: string | null;
    }
    openPeriode?:  {
      fom: string | null;
      extra: string;
    }
  };
  vedlegg ?: Array[any];
  trygdeordningnavn ?: string;
  medlemskap ?: string;
  forsikkringEllerRegistreringNr ?: string;
  annenInformasjon ?: string;
  usikkerDatoIndikator ?: string;
  jobbUnderAnsattEllerSelvstendig ?: string;
  navnFirma ?: string;
  typePeriode ?: string;
  adresseFirma ?: {
    postnummer: string | undefined;
    by: string | undefined;
    land: string;
    gate: string | undefined;
    region: string | undefined;
  };
  navnPaaInstitusjon ?: string;
  informasjonBarn ?:  {
    etternavn: string | null | undefined;
    fornavn: string | null | undefined;
    foedseldato: string | null | undefined;
    land: string;
  }
}

export interface Payload {
  ansattSelvstendigPerioder ?: Array<PayloadPeriod>;
  boPerioder ?: Array<PayloadPeriod>;
  barnepassPerioder ?: Array<PayloadPeriod>;
  frivilligPerioder ?: Array<PayloadPeriod>;
  forsvartjenestePerioder ?: Array<PayloadPeriod>;
  foedselspermisjonPerioder ?: Array<PayloadPeriod>;
  opplaeringPerioder ?: Array<PayloadPeriod>;
  arbeidsledigPerioder ?: Array<PayloadPeriod>;
  sykePerioder ?: Array<PayloadPeriod>;
  andrePerioder ?: Array<PayloadPeriod>;
}

export interface P4000Info {
  person: any,
  bank: any,
  stayAbroad: Array<Period>
}

export interface P4000PayloadInfo {
  personInfo: any,
  bankInfo: any,
  periodeInfo: Array<PayloadPeriod>
}
