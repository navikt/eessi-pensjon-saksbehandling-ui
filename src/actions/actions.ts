export interface Action<T> {
  type: string,
  payload?: T
}

export interface JustTypeAction {
  type: string
}

export interface PayloadError {
  error: any
}
