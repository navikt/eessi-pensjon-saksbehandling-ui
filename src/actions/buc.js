
const BUCLIST = {

  ONGOING: [
    {
      type: 'P_BUC_01',
      name: 'AldersPensjon',
      dateCreated: 'dd.mm.åå',
      countries: ['ZW', 'KR', 'SE', 'DK', 'CZ']
    },
    {
      type: 'P_BUC_02',
      name: 'UførePensjon',
      dateCreated: 'dd.mm.åå',
      countries: ['SE', 'DK', 'CZ'],
      merknader: ['foo', 'bar', 'baz']
    },
    {
      type: 'P_BUC_01',
      name: 'AldersPensjon',
      dateCreated: 'dd.mm.åå',
      countries: ['ZW', 'KR', 'SE', 'DK', 'CZ'],
      comments: ['foo', 'bar', 'baz']
    },
    {
      type: 'P_BUC_01',
      name: 'AldersPensjon',
      dateCreated: 'dd.mm.åå',
      countries: ['ZW', 'KR', 'SE', 'DK', 'CZ'],
      merknader: ['foo', 'bar', 'baz'],
      comments: ['foo', 'bar', 'baz']
    }
  ],
  OTHER: [
    {
      type: 'P_BUC_01',
      name: 'FIFOFA',
      dateCreated: 'dd.mm.åå',
      countries: ['RU', 'GB', 'CH']
    },
    {
      type: 'P_BUC_02',
      name: 'GRÅTASS',
      dateCreated: 'dd.mm.åå',
      countries: ['DE', 'TR', 'CA']
    }
  ]
}

const SEDS = [
  {
    name: 'P2000',
    status: 'sent',
    date: 'dd.mm.åå',
    country: 'Sverige',
    institution: 'Försäkringskassan'
  },
  {
    name: 'P3000SE',
    status: 'draft',
    date: 'dd.mm.åå',
    country: 'Sverige',
    institution: 'Försäkringskassan'
  },
  {
    name: 'P4000',
    status: 'received',
    date: 'dd.mm.åå',
    country: 'Sverige',
    institution: 'Försäkringskassan'
  },
  {
    name: 'P5000',
    status: 'foo',
    date: 'dd.mm.åå',
    country: 'Sverige',
    institution: 'Försäkringskassan'
  }
]

export const fetchBucList = async () => {
  return BUCLIST
}

export const fetchSedListForBuc = async (buc) => {
  return SEDS
}
