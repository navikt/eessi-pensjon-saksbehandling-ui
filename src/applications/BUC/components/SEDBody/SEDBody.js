import React from 'react'
// import SEDAttachments from '../SEDAttachments/SEDAttachments'

const SEDBody = ({ t, files }) => {
  return (
    <div className='a-buc-c-sedbody' />
  )
}

export default SEDBody

/*      <SEDAttachments t={t} files={files} /> */
