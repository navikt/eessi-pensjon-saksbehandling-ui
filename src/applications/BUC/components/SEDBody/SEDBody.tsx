import {
  createSavingAttachmentJob,
  resetSavingAttachmentJob,
  resetSedAttachments,
  sendAttachmentToSed
} from 'src/actions/buc'
import { sedAttachmentSorter } from 'src/applications/BUC/components/BUCUtils/BUCUtils'
import SEDAttachmentModal from 'src/applications/BUC/components/SEDAttachmentModal/SEDAttachmentModal'
import SEDAttachmentSender from 'src/applications/BUC/components/SEDAttachmentSender/SEDAttachmentSender'
import JoarkBrowser from 'src/components/JoarkBrowser/JoarkBrowser'
import {
  Buc,
  SavingAttachmentsJob,
  Sed,
  SEDAttachment,
  SEDAttachmentPayload,
  SEDAttachmentPayloadWithFile,
  SEDAttachments
} from 'src/declarations/buc'
import { BucPropType, SedPropType } from 'src/declarations/buc.pt'
import { JoarkBrowserItem, JoarkBrowserItems } from 'src/declarations/joark'
import { State } from 'src/declarations/reducers'
import _ from 'lodash'
import { Heading, Loader, Button, Box } from '@navikt/ds-react'
import PT from 'prop-types'
import { useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useDispatch, useSelector } from 'react-redux'

export interface SEDBodyProps {
  aktoerId: string
  buc: Buc
  canHaveAttachments: boolean
  initialAttachmentsSent?: boolean
  initialSeeAttachmentPanel?: boolean
  initialSendingAttachments?: boolean
  onAttachmentsSubmit?: (jbi: JoarkBrowserItems) => void
  onAttachmentsPanelOpen?: (o: boolean) => void
  sed: Sed
}

export interface SEDBodySelector {
  attachmentsError?: boolean
}

const mapState = (state: State): SEDBodySelector => ({
  attachmentsError: state.buc.attachmentsError
})

const SEDBody: React.FC<SEDBodyProps> = ({
  aktoerId,
  buc,
  canHaveAttachments,
  initialAttachmentsSent = false,
  initialSeeAttachmentPanel = false,
  initialSendingAttachments = false,
  onAttachmentsSubmit,
  sed
}: SEDBodyProps): JSX.Element => {
  const { attachmentsError }: SEDBodySelector = useSelector<State, SEDBodySelector>(mapState)
  const { t } = useTranslation()
  const dispatch = useDispatch()

  const convertSedAttachmentsToJoarkBrowserItems = (sedAttachments: SEDAttachments): JoarkBrowserItems | undefined => {
    if (!sedAttachments) {
      return undefined
    }
    return sedAttachments.map((att: SEDAttachment) => {
      return {
        key: att.id,
        type: 'sed',
        title: att.name,
        date: new Date(att.lastUpdate),
        hasSubrows: false,
        disabled: false,
        dokumentInfoId: att.documentId,
        journalpostId: '',
        openSubrows: false,
        tema: undefined,
        variant: undefined,
        visible: true,
        selected: false
      } as JoarkBrowserItem
    })
  }
  // initially, joark is empty
  const [_items, setItems] = useState<JoarkBrowserItems>(
    convertSedAttachmentsToJoarkBrowserItems(sed.attachments) || []
  )
  const [_sendingAttachments, setSendingAttachments] = useState<boolean>(initialSendingAttachments)
  const [_attachmentsSent, setAttachmentsSent] = useState<boolean>(initialAttachmentsSent)
  const [_attachmentsTableVisible, setAttachmentsTableVisible] = useState<boolean>(initialSeeAttachmentPanel)
  const [_currentPage, setCurrentPage] = useState<number>(1)

  const _sendAttachmentToSed = (params: SEDAttachmentPayloadWithFile, unsentAttachment: JoarkBrowserItem): void => {
    dispatch(sendAttachmentToSed(params, unsentAttachment))
  }

  const onAttachmentsPanelClose = (): void => {
    setAttachmentsTableVisible(false)
  }

  const onAttachmentsPanelOpen = (): void => {
    setAttachmentsTableVisible(true)
    setAttachmentsSent(false)
  }

  const onAttachmentsSubmitted = (): void => {
    setSendingAttachments(true)
    setAttachmentsTableVisible(false)
    const joarkItemsToUpload: JoarkBrowserItems = _.filter(_items, f => f.type === 'joark')
    dispatch(createSavingAttachmentJob(joarkItemsToUpload))
    if (_.isFunction(onAttachmentsSubmit)) {
      onAttachmentsSubmit(joarkItemsToUpload)
    }
  }

  const _onCancel = (): void => {
    setSendingAttachments(false)
    dispatch(resetSavingAttachmentJob())
  }

  const _onFinished = (): void => {
    setItems(_items.map(item => {
      item.type = 'sed'
      return item
    }))
    setAttachmentsSent(true)
  }

  const onJoarkAttachmentsChanged = (jbi: JoarkBrowserItems): void => {
    const sedOriginalAttachments: JoarkBrowserItems = _.filter(_items, (att) => att.type !== 'joark')
    const newAttachments = sedOriginalAttachments.concat(jbi).sort(sedAttachmentSorter)
    setItems(newAttachments)
  }

  const onRowViewDelete = (newItems: JoarkBrowserItems): void => {
    setItems(newItems)
  }

  const _onSaved = (savingAttachmentsJob: SavingAttachmentsJob): void => {
    const sedOriginalAttachments: JoarkBrowserItems = _.filter(_items, (att) => att.type === 'sed')
    const newAttachments: JoarkBrowserItems = sedOriginalAttachments
      .concat(savingAttachmentsJob.saved)
      .concat(savingAttachmentsJob.remaining)
      .sort(sedAttachmentSorter)
    setItems(newAttachments)
  }

  const onSedAttachmentsChanged = useCallback((sedAttachments: SEDAttachments) => {
    setItems(convertSedAttachmentsToJoarkBrowserItems(sedAttachments) || [])
  }, [setItems])

  useEffect(() => {
    // cleanup after attachments sent
    if (_sendingAttachments && _attachmentsSent) {
      setSendingAttachments(false)
      setAttachmentsTableVisible(false)
      dispatch(resetSedAttachments())
    }
    if (!_sendingAttachments && _attachmentsSent) {
      setAttachmentsSent(false)
      dispatch(resetSavingAttachmentJob())
    }
  }, [_attachmentsSent, dispatch, _sendingAttachments])

  useEffect(() => {
    onSedAttachmentsChanged(sed.attachments)
  }, [onSedAttachmentsChanged, sed])

  return (
    <Box>
      <Heading size='small'>
        {t('ui:attachments')}
      </Heading>
      {_.isEmpty(_items) && (
          <em>{t('ui:no-attachments')}</em>
        )
      }
      {!_.isEmpty(_items) && (
        <>
          <Box paddingBlock="8 0">
            <JoarkBrowser
              data-testid='a_buc_c_sedbody--attachments-id'
              existingItems={_items}
              mode='view'
              onRowViewDelete={onRowViewDelete}
              tableId={'viewsed-' + sed.id}
              currentPage={_currentPage}
              setCurrentPage={setCurrentPage}
            />
          </Box>
        </>
      )}
      <>
        {!_attachmentsSent && _.find(_items, (item) => item.type === 'joark') !== undefined && (
          <Box paddingBlock="4 0">
            <Button
              variant='primary'
              data-testid='a_buc_c_sedbody--upload-button-id'
              disabled={_sendingAttachments}
              onClick={onAttachmentsSubmitted}
            >
              {_sendingAttachments && <Loader />}
              {_sendingAttachments ? t('ui:uploading') : t('buc:form-submitSelectedAttachments')}
            </Button>
          </Box>
        )}
      </>
      {canHaveAttachments && (
        (_sendingAttachments || _attachmentsSent)
          ? (
              <Box marginBlock="4" width="100%">
                <>
                  <SEDAttachmentSender
                    attachmentsError={attachmentsError}
                    sendAttachmentToSed={_sendAttachmentToSed}
                    payload={{
                      aktoerId,
                      rinaId: buc.caseId,
                      rinaDokumentId: sed.id
                    } as SEDAttachmentPayload}
                    onCancel={_onCancel}
                    onFinished={_onFinished}
                    onSaved={_onSaved}
                  />
                </>
              </Box>
            )
          : (
            <>
              <Box paddingBlock="4 4">
                <Button
                  variant='secondary'
                  data-testid='a_buc_c_sedbody--show-table-button-id'
                  onClick={() => !_attachmentsTableVisible ? onAttachmentsPanelOpen() : onAttachmentsPanelClose()}
                >
                  {t(_attachmentsTableVisible ? 'ui:hideAttachments' : 'ui:showAttachments')}
                </Button>
              </Box>
            </>
            )
      )}
      <SEDAttachmentModal
        open={_attachmentsTableVisible}
        onModalClose={onAttachmentsPanelClose}
        onFinishedSelection={onJoarkAttachmentsChanged}
        sedAttachments={_items}
        tableId={'sedview' + sed.id + '-modal'}
      />
    </Box>
  )
}

SEDBody.propTypes = {
  aktoerId: PT.string.isRequired,
  buc: BucPropType.isRequired,
  canHaveAttachments: PT.bool.isRequired,
  initialAttachmentsSent: PT.bool,
  initialSeeAttachmentPanel: PT.bool,
  initialSendingAttachments: PT.bool,
  onAttachmentsSubmit: PT.func,
  onAttachmentsPanelOpen: PT.func,
  sed: SedPropType.isRequired
}

export default SEDBody
