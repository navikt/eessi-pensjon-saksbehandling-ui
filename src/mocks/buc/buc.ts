import { Institution } from 'declarations/buc'
import _ from 'lodash'
import mockBucs from 'mocks/buc/bucs'

export const mockBuc = (rinaCaseId: string) => _.find(mockBucs(), buc => buc.caseId === rinaCaseId)

export const mockParticipants = (rinaCaseId: string) => {
  const buc = mockBuc(rinaCaseId)
  return buc && buc.institusjon ? _.map(buc.institusjon, (i: Institution) => ({
    organisation: {
      countryCode: i.country,
      id: i.institution
    }
  })) : []
}
