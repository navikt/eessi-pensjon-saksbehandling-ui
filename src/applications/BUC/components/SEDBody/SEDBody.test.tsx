import { resetSavingAttachmentJob, resetSedAttachments } from 'actions/buc'
import { Buc } from 'declarations/buc'
import { render, screen } from '@testing-library/react'
import _ from 'lodash'
import mockBucs from 'mocks/buc/bucs'
import { stageSelector } from 'setupTests'
import SEDBody, { SEDBodyDiv, SEDBodyProps } from './SEDBody'

const defaultSelector = {
  attachmentsError: false
}

jest.mock('actions/buc', () => ({
  createSavingAttachmentJob: jest.fn(),
  resetSavingAttachmentJob: jest.fn(),
  resetSedAttachments: jest.fn(),
  sendAttachmentToSed: jest.fn()
}))

jest.mock('applications/BUC/components/SEDAttachmentModal/SEDAttachmentModal', () => {
  const joarkBrowserItems = require('mocks/joark/items').default
  return (props: any) => (
    <div data-testid='mock-sedattachmentmodal'>
      <button id='onFinishedSelection' onClick={() => props.onFinishedSelection(joarkBrowserItems)}>click</button>
    </div>
  )
})

jest.mock('components/JoarkBrowser/JoarkBrowser', () => () => (
  <div data-testid='mock-joarkbrowser' />
))

describe('applications/BUC/components/SEDBody/SEDBody', () => {
  let wrapper: any
  const buc: Buc = mockBucs()[0]
  const initialMockProps: SEDBodyProps = {
    aktoerId: '123',
    buc,
    canHaveAttachments: true,
    initialAttachmentsSent: false,
    initialSeeAttachmentPanel: false,
    initialSendingAttachments: false,
    onAttachmentsSubmit: jest.fn(),
    onAttachmentsPanelOpen: jest.fn(),
    sed: buc.seds![0]
  }

  beforeEach(() => {
    stageSelector(defaultSelector, {})
    wrapper = render(<SEDBody {...initialMockProps} />)
  })

  afterEach(() => {
    wrapper.unmount()
  })

  it('Render: match snapshot', () => {
    const { container } = render(<SEDBody {...initialMockProps} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  it('Render: has proper HTML structure', () => {
    expect(wrapper.exists(SEDBodyDiv)).toBeTruthy()
  })

  it('Render: shows JoarkBrowser only if items are not empty', () => {
    expect(_.isEmpty(initialMockProps.sed.attachments)).toBeFalsy()
    expect(screen.getByTestId('a_buc_c_sedbody--attachments-id')).toBeTruthy()
    const newMockProps = {
      ...initialMockProps,
      sed: {
        ...initialMockProps.sed,
        attachments: []
      }
    }
    wrapper = render(<SEDBody {...newMockProps} />)
    expect(screen.getByTestId('a_buc_c_sedbody--attachments-id')).toBeFalsy()
  })

  it('Render: shows SEDAttachmentSender', () => {
    wrapper = render(<SEDBody {...initialMockProps} />)
    expect(wrapper.exists('SEDAttachmentSender')).toBeFalsy()
    wrapper = render(<SEDBody {...initialMockProps} initialSendingAttachments />)
    expect(wrapper.exists('SEDAttachmentSender')).toBeTruthy()
  })

  it('Render: shows upload button if there are joark attachments to upload', () => {
    wrapper = render(<SEDBody {...initialMockProps} initialSeeAttachmentPanel />)
    expect(screen.getByTestId('a_buc_c_sedbody--upload-button-id')).toBeFalsy()
    wrapper.find('#onFinishedSelection').hostNodes().simulate('click')
    expect(screen.getByTestId('a_buc_c_sedbody--upload-button-id')).toBeTruthy()
  })

  it('Render: shows modal open button if SED can have attachments', () => {
    expect(screen.getByTestId('mock-sedattachmentmodal\']')).toBeFalsy()
    wrapper.find('[data-testid=\'a_buc_c_sedbody--show-table-button-id').hostNodes().simulate('click')
    expect(screen.getByTestId('mock-sedattachmentmodal\']')).toBeTruthy()
  })

  it('UseEffect: cleanup after attachments sent', () => {
    (resetSedAttachments as jest.Mock).mockReset()
    wrapper = render(<SEDBody {...initialMockProps} initialAttachmentsSent initialSendingAttachments />)
    expect(resetSedAttachments).toHaveBeenCalled()
  })

  it('UseEffect: cleanup after attachments sent, part 2', () => {
    (resetSavingAttachmentJob as jest.Mock).mockReset()
    wrapper = render(<SEDBody {...initialMockProps} initialAttachmentsSent initialSendingAttachments={false} />)
    expect(resetSavingAttachmentJob).toHaveBeenCalled()
  })
})
