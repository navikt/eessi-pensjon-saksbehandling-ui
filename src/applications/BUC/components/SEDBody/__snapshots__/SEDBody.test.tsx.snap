// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applications/BUC/components/SEDBody/SEDBody Render: match snapshot 1`] = `
.c1 {
  margin-bottom: 1rem;
}

.c2 {
  margin-bottom: 2rem;
}

ui-fss.g19[id="c0"] {
  content: "c1,c2,";
}

<div
  class=""
>
  <div
    class="c0 c1"
  />
  <h1
    class="navds-heading navds-heading--small"
  >
    ui:attachments
  </h1>
  <div
    class="c0 c2"
    size="2"
  />
  <div
    data-testid="mock-joarkbrowser"
  />
  <div
    class="c0 c1"
  />
  <div
    class="c0 c1"
  />
  <button
    class="navds-button navds-button--secondary navds-button--medium"
    data-testid="a_buc_c_sedbody--show-table-button-id"
  >
    <span
      aria-live="polite"
      class="navds-label"
    >
      ui:showAttachments
    </span>
  </button>
  <div
    class="c0 c1"
  />
  <div
    data-testid="mock-sedattachmentmodal"
  >
    <button
      id="onFinishedSelection"
    >
      click
    </button>
  </div>
</div>
`;
